### DISCUSSION TITLE
ISA Compatibility

### DISCUSSION THREAD

**Discussion by fbacall** [seek4science/seek#517] [Q&A]
SEEK's ISA model and the ISATab/JSON formats are not quite compatible. 

We should think about how to align them so we could make use of the various ISA tools to assist with converting between formats and exporting to repositories.

**Comment by floradanna** [comment]
Regarding the question, I would need a clarification about the type of metadata we are discussing.

1. I assume you are refering strictly to any Object and Attribute defined by the ISA specifications and formalized in the ISA JSON schema 1.0 ?

So I understand your question as: for the Study Object, which are the fields which should be minimally supplied?

If that's the case, then for most objects, the required information is Name and Identifier (if not autogenerated by the API). 

For specific objects such as ISA.Protocol, ISA.Factor, ISA.Assay , the associated 'types' must be provided. so `ISA.Protocol.type` and `ISA.StudyFactor.type`

For ISA.Assay, `ISA.Assay.measurement_type` must be provided, and `ISA.Assay.technology_type` is strongly recommended but not mandatory.

2. If "metadata" refers to descriptors outside the ISA syntax 

I am talking about things such as 'Disease', 'Instrument Temperature', 'Organism Part', then one may define ISA.Configurations/metadata profiles but such requirements are not to be found in the ISA schema itself

Let me know if this helps and if more information is needed.

Also, let me know about any mapping FAIRDOM-SEEK is making so I can review it and possibly advise.

**Comment by floradanna** [comment]
Additional info about ISA.
- The documentation is  available from the ISA-Tab specifications: https://isa-specs.readthedocs.io/en/latest/isatab.html. Alternately, in what from, we would want to have this information?

- Descriptors outside ISA syntax are configurable, and could be defined based on metadata elements require by a EBI repository (for instance, ENA), but only to some extend. If you want to use some of the ISA conversion functions, you'd need to use the configurations we have defined and support for the deposition paths. If you choose to develop / redevelop the code, please do let the us (the ISA Team) known so we make sure we don't have incompatible implementation of the same checklists.

**Comment by floradanna** [comment]
@proccaserra
1) can 2 assays in the same study have 2 different measurement type?
2) must the number of "sample entries" be the same of "source entries"? Or the number of "sample entries" in an assay can be bigger or smaller than the number of "source entries" in the study?

**Comment by proccaserra** [comment]
@floradanna 
1. Yes, absolutely. this is why ISA was designed in the first place, to allow multi-omics, multi-modality data acquisition studies to be reported.

STUDY ASSAYS |   |   |   |  
-- | -- | -- | -- | --
Study Assay File Name | a_non-polar_GC_negative_assay.txt | a_illumina_sequencing_assay.txt | a_oxfordnanopore_sequencing_assay.txt | a_both_combined_sequencing_assay.txt
Study Assay Measurement Type | metabolite profiling | transcription profiling | transcription profiling | transcription profiling
etc... |   |   |   |  
  |   |   |   |  
Study Assay Technology Type | mass spectrometry | nucleic acid sequencing | nucleic acid sequencing | nucleic acid sequencing


2. Source / Sample and Assays:
Any `Sample` used in an Assay table (if using ISA-Tab serialisation), *must* be declared. Failing that will cause the validator to raise an error. 

- the number of Samples  may be bigger than the number of Sources (e.g. multiple samples collected from a same Source, aliquoting in multiple subsamples, fractionation...)
- the number of Samples may be smaller than the number of Sources in the case of pooling events (5 mice pr group but hippocampus are pooled to performed some cell sorting)

Do you have situations where you have 'Samples' used for a data acquisition which have no declared Sources?

**Comment by floradanna** [comment]
When I used the ISA creator, I created a study with 2 rows in source and 4 rows as samples (2 samples from each source row); however I got error and I had to duplicate the sources in order to fill in 4 rows as well. So, I was wondering how this would work for pooled samples instead.

Also, in the assay file, can I only list a subset of the samples that are declared in the study file? For instance, 2 samples (out of 4) are used in assayA and the other 2 samples (out of 4) are used in assayB, in the same study.

**Comment by proccaserra** [comment]
@floradanna can you share the file you created + configuration you've used?

**Comment by floradanna** [comment]
I will have to re-create it because I don't always manage to save the exported tabs. I will let you know when I have it done. Thank you!

**Comment by floradanna** [comment]
@proccaserra I tested my own questions by using the ISA creator, however it is pity that I don't manage to see nor select data file to continue (image 7 of the pdf in attachment). There is also a problem with a SRA centre name that I don't know how to fix, therefore I cannot covert to ISA tab.

Anyway, I think I have the information I needed it for now. Thanks!


[ISAcreator.pdf](https://github.com/seek4science/seek/files/5781808/ISAcreator.pdf)

**Comment by floradanna** [comment]
@proccaserra Hi Philippe, I was wondering if you can have more than one "sample collection" protocol in ISA. From what I can see, it seems that there should be only ONE single "sample collection" protocol valid for all assay types in a study. Please, let me know if I am wrong. Thanks.

**Comment by proccaserra** [comment]
@floradanna yes this is the case. we are documenting this behaviour in a notebook. If this is a problem, could you document what information or what situation this becomes an issue ?

On top of my head, it can be possibly an issue when collecting different sample types from patients (blood, lung fluid, muscle biopsy) or various meta genomics samples.

a work around is to declare one or more parameters attached to the sample collection protocol which would give more flexibility.

I will get back to the other comment shortly . all the best.

**Comment by proccaserra** [comment]
@floradanna: 
1. can you share the ISA-Tab you have created with ISAcreator ?
2. Which version of ISA creator did you use and which with Configuration?
3. Image2: this is the expected  behaviour: an ISA record correspond to a the path of the graph projected in a 2d table so a 'branching' even where 1 source gives yield to 2 samples is represented by 2 rows in an ISA table. ISAcreator (java code) makes no assumptions based on the strings found in Material Nodes.
4. Image4: there seems to be a naming error for the second "Labeled Extract Name". This may sound contracdictory to my point above but only relates to the 'naming convention' you seem to have applied in the example. In the current example, the ISA parser (java & python) will assume that samples 'source1_a' and 'source2_a' are pooled into one extract named 'source1_b_extract', which is itself input to a labelling process which will generate 2 distinct labeled extracts , named respectively 'source1_b_labeled' and 'source2_b_labeled'. Is this the intent?
5. image 7: ISAcreator being no longer maintained, it could be a consequence of class deprecation in some of the libraries we relied on or settings to access folder information. Which OS are you running ISAcreator on?
Am i right assuming you are looking for fastq files for the sequencing application?
I'll be testing again on my end but like I said, ISAcreator is no longer supported as we focus on the python api.
6. Image 8: This is specific to the SRA conversion. It is (was) at the time of development a requirement to obtain a valid SRA Center Name in order to perform the conversion. This is supplied via a Comment field. Hence my question about which configuration you have been using?

hope this helps
bw
P

> @proccaserra I tested my own questions by using the ISA creator, however it is pity that I don't manage to see nor select data file to continue (image 7 of the pdf in attachment). There is also a problem with a SRA centre name that I don't know how to fix, therefore I cannot covert to ISA tab.
> 
> Anyway, I think I have the information I needed it for now. Thanks!
> 
> [ISAcreator.pdf](https://github.com/seek4science/seek/files/5781808/ISAcreator.pdf)

**Comment by floradanna** [comment]
1. I have never managed to save an ISA-TAB with ISA creator Light (ISAcreator-1.7.11) because of errors. I only managed to save it by using the MetaboLights ISA Creator (see attachment).
2. ISAcreator-1.7.11; I am sure about the configuration, I think this one "isaconfig-default_v2015-07-02".
3. OK, thank you.
4. Image4: I made a naming mistake in the second extract name (it should have been source2_b_extract, not source1_b_extract). I didn't intent to pool sources nor samples. I wanted to use source1_a and source2_a in Transcriptomic assay (image 3), and source1_b and source2_b in Metabolomic assay (image 4).
5. image 7: macOS Catalina Version 10.15.7
6. Image 8: see point 1.

[Document_archive.zip](https://github.com/seek4science/seek/files/5801323/Document_archive.zip)

@proccaserra  Thank you very much for your answers.

**Comment by floradanna** [comment]
@proccaserra 
Indeed, it could be an issue every time different sample types (blood/urine, root/leaf, cells/supernatant, tree/soil etc...) from a source (human, animal, plant, cell culture, environment etc...) need to be taken.
However, I would keep it as it is now: one single "sampling" protocol for the all study, otherwise everything could get more complicated. Also, this kind of issue can happen for every protocol, not only for sampling. For instance, you could extract RNA by Trizol for half of the samples and by Column for the other half (if extraction protocol is your experimental factor).

Therefore, I would keep it as it is. A simple solution is the the user will need to add more "characteristics" to describe samples, some of which apply to only a subset of samples and the rest of samples will have NA as value.

Please, let us know if you are going to change this in ISA anyway.

> @floradanna yes this is the case. we are documenting this behaviour in a notebook. If this is a problem, could you document what information or what situation this becomes an issue ?
> 
> On top of my head, it can be possibly an issue when collecting different sample types from patients (blood, lung fluid, muscle biopsy) or various meta genomics samples.
> 
> a work around is to declare one or more parameters attached to the sample collection protocol which would give more flexibility.
> 
> I will get back to the other comment shortly . all the best.

**Comment by sorenwacker** [comment]
Reviving an old thread with a concrete data point, in case it is useful for the alignment question.

We maintain the CropXR metadata model as SEEK definition files — sample types plus extended metadata types — and have just modelled it into an external schema representation (a metaseed profile) for our phenotyping, sequencing and metabolomics streams. Four places where the SEEK model did not map cleanly onto an ISA-shaped tree, which is roughly the same friction that makes ISATab conversion awkward.

**1. `Input` is many-valued, so the sample chain is a graph, not a containment tree.**

Sample types attach at four levels (`study source`, `study sample`, `assay - material`, `assay - data file`) and chain by an `Input` attribute of type `Registered Sample List`. Because a row may select several upstream rows, the lineage is a DAG. Any target model that expresses parentage by containment loses that; we had to keep `Input` as an explicit cross-reference alongside the nesting, so a row's containing record and its actual upstream inputs are two different things. This is the same branching/pooling case discussed above, but it shows up structurally rather than only in the 2D projection.

**2. Nested extended metadata links resolve by numeric instance id.**

Types with `supported_type: ExtendedMetadata` are embedded in a Study through a `Linked Extended Metadata` attribute. The link is made by the numeric id of the nested type on the target instance, so the same definition file means different things on two instances, and definitions are not portable without a build step that rewrites ids. A stable name-based reference would make the definitions instance-independent and much easier to convert.

**3. Row identity and ISA role are carried by two different mechanisms.**

Exactly one attribute per sample type sets `"title": true`, and that column is the row's identity — it is what other tables show when selecting inputs. Separately, every attribute carries an `isaTag` (`source`, `sample`, `other_material`, `data_file`, `protocol`, `parameter_value`, …), and it is the tag, not the name, that determines how a value is emitted on ISA export. Two attributes with the same name and different tags are not interchangeable. So the ISA node name and the SEEK display column are declared independently and can disagree. Consumers have to read both to reconstruct an ISA graph.

**4. The condensed assay types collapse two ISA levels into one table.**

For each stream we offer both a full form (an `assay - material` table feeding a separate `assay - data file` table) and a condensed form that merges protocol and file fields into a single `assay - data file` table. The condensed form records the measurement description once per file rather than once per assay, which has no clean ISA counterpart — on export it either duplicates the material node per file or drops it.

One smaller thing: an Assay record that carries only extended metadata has no identifier attribute of its own, so the record that several sample tables hang off has no stable key inside the definitions themselves.

None of this is a bug report — the model works well for data entry, which is what it is for. But if the alignment work is still live, these are the four points where we had to make a judgement call rather than a mechanical translation, and where a small amount of extra declaration in the definition format (name-based nested-type references, an explicit identity attribute) would remove the ambiguity.

**Comment by sorenwacker** [comment]
Following up on my comment above. We went on to build an abstract model of the SEEK definition format itself — entities for sample type, sample attribute, extended metadata type, extended metadata attribute and controlled vocabulary — and round-tripped our 11 sample types and 8 extended metadata types through it. Four things came out of that which are relevant here.

**First, a correction to my own approach.** I built the model by reading our definition files, and only afterwards found `lib/seek/isa_templates/template_attributes_schema.json`. That schema is genuinely useful and I had missed it. Validating against it immediately showed my model was too narrow: I had inferred a `dataType` enum of 8 values from what our files happen to use, where the schema declares 18 (`Date time`, `Date`, `Web link`, `Email address`, `Boolean`, `Registered Sample`, `Controlled Vocabulary List`, `URI`, `Registered SOP`, `Registered Strain` were all absent from mine). Anyone deriving a model from real-world templates rather than from the schema will under-approximate the format the same way. It would help if the template documentation linked to that file.

**Second, the schema covers attributes but not the template itself.** There is no published counterpart for the `metadata` block, so the keys we depend on — `level`, `group`, `group_order`, `temporary_name`, `organism`, `version`, and `isa_measurement_type` / `isa_technology_type` — are discoverable only by reading `template_extractor.rb` or the seed templates. We found the two ISA type keys by noticing them in our own files, not from any documentation. A schema for the template level would close that.

**Third, the attribute schema and the model validation disagree about `isaTag`.** The schema declares it as

```json
"isaTag": { "anyOf": [ { "type": "string", "enum": [...] }, { "type": "null" } ] }
```

so a null `isaTag` passes schema validation. But `Template#validate_template_attributes` rejects it:

```ruby
if attributes_with_empty_isa_tag.any?
  errors.add("[#{:template_attributes}]:", "Attribute '#{attribute.title}' is missing an ISA tag")
end
```

An external tool that validates against the published schema and gets a clean result will still have its upload refused. Either the schema should require `isaTag`, or it should be documented as necessary but not sufficient. We hit this as a `422` at upload time before we reimplemented the check ourselves.

**Fourth, the cross-attribute constraints are not expressible in the published schema.** `test_tag_occurences`, the requirement that exactly one attribute sets `title`, and the limit of one `Registered Sample List` attribute per template are all enforced in Ruby only. Draft-04 JSON Schema cannot express most of them, so this is not a criticism of the schema file — but the practical effect is that every external toolchain reimplements the same four rules from prose, and they drift. Publishing them in any machine-readable form, even a plain rules list alongside the schema, would remove that duplication.

None of this blocks us; we validate against our own copy of the rules. But three of the four points are cases where the information exists in the codebase and not in a form external tooling can consume, which seems close to the original question in this thread about aligning with the ISA tooling ecosystem.

**Comment by sorenwacker** [comment]
Short follow-up: I checked the two points above against a running instance (`fairdom/seek:1.18.1`) rather than by reading source, and one of my numbers was too generous.

**The `isaTag` disagreement is real.** Posting a template whose attribute carries no ISA tag:

```
HTTP 422
{"[template_attributes]:":["Attribute 'probe_col' is missing an ISA tag"]}
```

The same attribute with a valid tag returns 200. So an attribute that satisfies `template_attributes_schema.json` — which declares `isaTag` as `anyOf: [enum, null]` — is refused by the server. The cross-attribute rule behaves the same way, with a good error message:

```
HTTP 422
{"[template_attributes]:":["The 'protocol' ISA Tag was used in these attributes => [\"a\", \"b\"].
  This ISA tag is not allowed to be used more then once!"]}
```

Both rules are worth having; the point is only that neither is discoverable from the published schema, so external tooling finds them by being rejected.

**Correction to my previous comment.** I said the schema declares 18 `dataType` values against the 8 I had inferred. Querying `/sample_attribute_types.json` on the instance returns **28**. The published schema is a strict subset — everything in it exists on the instance, but ten types do not appear in it at all:

`ChEBI`, `DOI`, `ECN`, `InChI`, `NCBI ID`, `MetaNetX chemical`, `MetaNetX reaction`, `MetaNetX compartment`, and — most relevant to this thread — `Linked Extended Metadata` and `Linked Extended Metadata (multiple)`.

So there are three different answers to "what data types may a template attribute have", depending on whether you read a project's own documentation, the schema file in the repository, or the instance you are deploying to. We had all three disagree in one afternoon.

`/isa_tags.json` by contrast returns exactly the 11 values in the schema enum, so the tag vocabulary is consistent. That let us settle a discrepancy in our own documentation, which had been carrying a twelfth tag, `observation_unit_characteristic`, that does not exist on the instance and never did. That one was our error, not SEEK's.

The concrete ask stays as before: if the schema file is intended as the contract for template attributes, having it generated from the same source as `sample_attribute_types` — or simply noting in the docs that it may lag the instance and that `/sample_attribute_types.json` is authoritative — would stop external tools encoding a stale enum.
